File: exercises/cs-core/traceable-spanning-forest-lab/python/tests/test_lab.py
from __future__ import annotations

import subprocess
import sys
import unittest

from traceable_spanning_forest_lab import (
    DisjointSet,
    SpanningGraph,
    compare_spanning_forests,
    kruskal_forest,
    lazy_prim_forest,
    minimum_spanning_tree,
    rejected_cycle_edges,
)


class DisjointSetTests(unittest.TestCase):
    def test_zero_single_and_bounds(self) -> None:
        self.assertEqual(DisjointSet(0).groups(), ())
        dsu = DisjointSet(1)
        self.assertEqual(dsu.find(0).visits, 1)
        with self.assertRaises(IndexError):
            dsu.find(1)

    def test_union_tie_and_repeated_merge(self) -> None:
        dsu = DisjointSet(4)
        self.assertEqual(dsu.union(1, 0).root, 0)
        self.assertEqual(dsu.union(3, 2).root, 2)
        self.assertEqual(dsu.union(2, 0).root, 0)
        self.assertFalse(dsu.union(3, 1).merged)
        self.assertEqual(dsu.component_count, 1)

    def test_full_compression_and_groups(self) -> None:
        dsu = DisjointSet(7)
        for edge in ((0, 1), (2, 3), (0, 2), (4, 5), (5, 6), (3, 6)):
            dsu.union(*edge)
        self.assertEqual(dsu.parents, (0, 0, 0, 0, 0, 4, 4))
        trace = dsu.find(5)
        self.assertEqual((trace.path, trace.visits, trace.compressions), ((5, 4, 0), 3, 1))
        self.assertEqual(dsu.groups()[0].members, tuple(range(7)))


class ForestTests(unittest.TestCase):
    def setUp(self) -> None:
        self.graph = SpanningGraph(7, (
            (0, 1, 4), (0, 2, 1), (1, 2, 2), (1, 3, 5),
            (2, 3, 3), (2, 4, 6), (3, 4, 2), (5, 6, -1),
        ))

    def test_graph_validation(self) -> None:
        for edges in (((0, 0, 1),), ((0, 2, 1),), ((0, 1, 1), (1, 0, 2))):
            with self.assertRaises(ValueError):
                SpanningGraph(2, edges)
        self.assertEqual(SpanningGraph(2, ((0, 1, 0),)).edges[0].weight, 0)

    def test_kruskal_disconnected_and_cycles(self) -> None:
        forest = kruskal_forest(self.graph)
        self.assertEqual((forest.total_weight, forest.component_count, len(forest.edges)), (7, 2, 5))
        self.assertEqual(len(rejected_cycle_edges(self.graph)), 3)
        with self.assertRaises(ValueError):
            minimum_spanning_tree(self.graph)

    def test_input_order_and_equal_weights_are_deterministic(self) -> None:
        first = SpanningGraph(4, ((2, 3, 1), (0, 2, 1), (0, 1, 1), (1, 3, 1)))
        second = SpanningGraph(4, reversed(first.edges))
        self.assertEqual(kruskal_forest(first).edges, kruskal_forest(second).edges)

    def test_prim_counts_and_comparison(self) -> None:
        trace = lazy_prim_forest(self.graph)
        self.assertEqual(trace.component_starts, (0, 5))
        self.assertEqual((trace.edge_scans, trace.queue_pushes, trace.stale_pops, trace.max_frontier), (16, 8, 3, 4))
        self.assertTrue(compare_spanning_forests(self.graph).matching)

    def test_empty_graph_and_isolated_vertices(self) -> None:
        trace = lazy_prim_forest(SpanningGraph(3, ()))
        self.assertEqual((trace.component_starts, trace.forest.component_count), ((0, 1, 2), 3))
        self.assertEqual(trace.forest.edges, ())

    def test_total_weight_overflow(self) -> None:
        graph = SpanningGraph(3, ((0, 1, (1 << 63) - 1), (1, 2, 1)))
        with self.assertRaises(OverflowError):
            kruskal_forest(graph)
        with self.assertRaises(OverflowError):
            lazy_prim_forest(graph)


class ReportingTests(unittest.TestCase):
    def test_modes_and_unknown_mode(self) -> None:
        for mode, title in (("dsu", "可追踪并查集"), ("kruskal", "Kruskal 最小生成森林"), ("prim", "Lazy Prim 最小生成森林")):
            result = subprocess.run([sys.executable, "-m", "traceable_spanning_forest_lab", mode], text=True, capture_output=True, check=True)
            self.assertTrue(result.stdout.startswith(title))
            self.assertEqual(result.stderr, "")
        failed = subprocess.run([sys.executable, "-m", "traceable_spanning_forest_lab", "bad"], text=True, capture_output=True, check=False)
        self.assertEqual((failed.returncode, failed.stdout), (2, ""))
        self.assertIn("unknown mode", failed.stderr)


if __name__ == "__main__":
    unittest.main()

File: exercises/cs-core/traceable-spanning-forest-lab/cpp/tests/test_lab.cpp
#include "traceable_spanning_forest_lab/lab.hpp"
#include <cassert>
#include <cstdint>
#include <limits>
#include <stdexcept>
using namespace traceable_spanning_forest_lab;
int main() {
  DisjointSet dsu(7);
  for (const auto& edge : {std::pair{0U,1U},std::pair{2U,3U},std::pair{0U,2U},std::pair{4U,5U},std::pair{5U,6U},std::pair{3U,6U}}) assert(dsu.unite(edge.first,edge.second).merged);
  assert((dsu.parents() == std::vector<std::size_t>{0,0,0,0,0,4,4}));
  const auto found=dsu.find(5); assert(found.visits==3 && found.compressions==1 && found.root==0);
  bool range_failed=false; try { dsu.find(7); } catch(const std::out_of_range&) { range_failed=true; } assert(range_failed);
  const auto graph=sample_spanning_graph(); const auto kruskal=kruskal_forest(graph); const auto prim=lazy_prim_forest(graph);
  assert(kruskal.total_weight==7 && kruskal.component_count==2 && kruskal.edges.size()==5);
  assert(prim.edge_scans==16 && prim.queue_pushes==8 && prim.stale_pops==3 && prim.max_frontier==4);
  assert(compare_spanning_forests(graph).matching);
  bool tree_failed=false; try { minimum_spanning_tree(graph); } catch(const std::invalid_argument&) { tree_failed=true; } assert(tree_failed);
  bool duplicate_failed=false; try { SpanningGraph bad(2,{{0,1,1},{1,0,2}}); } catch(const std::invalid_argument&) { duplicate_failed=true; } assert(duplicate_failed);
  bool overflow_failed=false; try { kruskal_forest(SpanningGraph(3,{{0,1,std::numeric_limits<std::int64_t>::max()},{1,2,1}})); } catch(const std::overflow_error&) { overflow_failed=true; } assert(overflow_failed);
  assert(build_dsu_report().starts_with("可追踪并查集")); assert(build_kruskal_report().starts_with("Kruskal")); assert(build_prim_report().starts_with("Lazy Prim"));
}
